datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,1739 @@
+///
+/// This method will return a Vec [3, 0, 1, 2]
+fn expected_expr_positions(
+    current: &[Arc<dyn PhysicalExpr>],
+    expected: &[Arc<dyn PhysicalExpr>],
+) -> Option<Vec<usize>> {
+    let mut indexes: Vec<usize> = vec![];
+    let mut current = current.to_vec();
+    for expr in expected.iter() {
+        // Find the position of the expected expr in the current expressions
+        if let Some(expected_position) = current.iter().position(|e| e.eq(expr)) {
+            current[expected_position] = Arc::new(NoOp::new());
+            indexes.push(expected_position);
+        } else {
+            return None;
+        }
+    }
+    Some(indexes)
+}
+
+fn extract_join_keys(on: &[(Column, Column)]) -> JoinKeyPairs {
+    let (left_keys, right_keys) = on
+        .iter()
+        .map(|(l, r)| {
+            (
+                Arc::new(l.clone()) as Arc<dyn PhysicalExpr>,
+                Arc::new(r.clone()) as Arc<dyn PhysicalExpr>,
+            )
+        })
+        .unzip();
+    JoinKeyPairs {
+        left_keys,
+        right_keys,
+    }
+}
+
+fn new_join_conditions(
+    new_left_keys: &[Arc<dyn PhysicalExpr>],
+    new_right_keys: &[Arc<dyn PhysicalExpr>],
+) -> Vec<(Column, Column)> {
+    let new_join_on = new_left_keys
+        .iter()
+        .zip(new_right_keys.iter())
+        .map(|(l_key, r_key)| {
+            (
+                l_key.as_any().downcast_ref::<Column>().unwrap().clone(),
+                r_key.as_any().downcast_ref::<Column>().unwrap().clone(),
+            )
+        })
+        .collect::<Vec<_>>();
+    new_join_on
+}
+
+fn ensure_distribution_and_ordering(
+    plan: Arc<dyn crate::physical_plan::ExecutionPlan>,
+    target_partitions: usize,
+) -> Arc<dyn crate::physical_plan::ExecutionPlan> {
+    if plan.children().is_empty() {
+        return plan;
+    }
+    let required_input_distributions = plan.required_input_distribution();
+    let required_input_orderings = plan.required_input_ordering();
+    let children: Vec<Arc<dyn ExecutionPlan>> = plan.children();
+    assert_eq!(children.len(), required_input_distributions.len());
+    assert_eq!(children.len(), required_input_orderings.len());
+
+    // Add RepartitionExec to guarantee output partitioning
+    let children = children
+        .into_iter()
+        .zip(required_input_distributions.into_iter())
+        .map(|(child, required)| {
+            if child
+                .output_partitioning()
+                .satisfy(required.clone(), || child.equivalence_properties())
+            {
+                child
+            } else {
+                let new_child: Arc<dyn ExecutionPlan> = match required {
+                    Distribution::SinglePartition
+                        if child.output_partitioning().partition_count() > 1 =>
+                    {
+                        Arc::new(CoalescePartitionsExec::new(child.clone()))
+                    }
+                    _ => {
```

Comment:
Here, we are sure that the `RepartitionExec` in the plan tree should satisfy the `required` distribution so that we don't need to worry about duplicate `RepartitionExec` be added.

Reply:
Isn't that case covered by `if child.output_partitioning().satisfy()` check above?